# This script aims to compare the relatedness matrices from different sources. It currently compares:
# - Relatedness based on pedigree
# - Relatedness based on SNP data, centered GRM, generated by GEMMA or KING
# - Relatedness based on kmers, generated by kgwasflow/kmersgwas/EMMA

source('quant_gen/prepare_pedigrees.R')
library(VariantAnnotation)
library(tidyverse)

vcf_file <- 'sequencing/gwas/filtered.vcf.gz'
if (!file.exists(paste0(vcf_file, '.tbi'))) indexVcf(vcf_file)

### Data prep
sampling1 <- data.table::fread('sequencing/sample_lists/DNA_WGS_300samples_long_list.csv')
sampling2 <- data.table::fread('data/NovaSeq_sample_list.csv') %>%
  dplyr::select(sample_id = Name, sample_name = `Filename Prefix`)
sampling <- inner_join(sampling1, sampling2, by = 'sample_id') %>%
  dplyr::select(fish_id, sample_id, sample_name) %>%
  mutate(
    fish_id = tolower(fish_id),
    # adjust these names, for samples that were sequenced twice.
    sample_name = case_when(
      sample_name == 'NS.2145.001.IDT_i7_77---IDT_i5_77.289' ~ 'sample_289',
      sample_name == 'NS.2145.001.IDT_i7_144---IDT_i5_144.323' ~ 'sample_323',
      TRUE ~ sample_name
    )
  )

### Visualize the estimated relatedness matrix
gplots::heatmap.2(
  data.table::fread('sequencing/gwas/gemma_output/kinship.cXX.txt') %>% as.matrix(),
  symm = TRUE, trace = 'none'
)

# Get the order of samples from the VCF, since that's the order GEMMA uses
#
sample_names <- readVcf(vcf_file, param = ScanVcfParam(which = GRanges("NC_024331.1", IRanges(24115677)))) %>%
  colData() %>% rownames()

# Load the SNP GRM from GEMMA
SNP_relatedness <- data.table::fread('sequencing/gwas/gemma_output/for_kinship_comparison.cXX.txt') %>%
  as.data.frame() %>%
  `dimnames<-`(list(sample_names, sample_names)) %>%
  rownames_to_column(var = 'id1') %>%
  pivot_longer(-id1, names_to = 'id2', values_to = 'r_SNP') %>%
  left_join(sampling, c('id1' = 'sample_name')) %>%
  left_join(sampling, c('id2' = 'sample_name'), suffix = c('1', '2'))

KING_relatedness <- data.table::fread('sequencing/gwas/gemma_output/king_kinship_for_comparison.king') %>%
  as.data.frame() %>%
  `dimnames<-`(list(sample_names, sample_names)) %>%
  rownames_to_column(var = 'id1') %>%
  pivot_longer(-id1, names_to = 'id2', values_to = 'r_SNP_KING')

# Get the order of listed samples, since that is what kgwasflow uses
sample_names_kgwas <- data.table::fread('sequencing/kmer_gwas/config/samples.tsv')$sample_name %>%
  # remove the duplicates (because some samples were sequenced twice)
  unique()

# Load the kmer GRM
kmer_relatedness <- data.table::fread('sequencing/kmer_gwas/tmp_files/kmers_table.kinship') %>%
  as.data.frame() %>%
  `dimnames<-`(list(sample_names_kgwas, sample_names_kgwas)) %>%
  rownames_to_column(var = 'fish_id1') %>%
  pivot_longer(-fish_id1, names_to = 'fish_id2', values_to = 'r_kmer')

fish <- unique(SNP_relatedness$fish_id1)
pedigree_relatedness <- A[fish, fish] %>%
  as.data.frame() %>%
  rownames_to_column(var = 'fish_id1') %>%
  pivot_longer(-fish_id1, names_to = 'fish_id2', values_to = 'r_pedigree')

rel_joined <- inner_join(SNP_relatedness, pedigree_relatedness, by = c('fish_id1', 'fish_id2')) %>%
  full_join(KING_relatedness, by = c('id1', 'id2')) %>%
  full_join(kmer_relatedness, by = c('fish_id1', 'fish_id2'))

(
  comp <- ggplot(rel_joined, aes(r_SNP, r_pedigree)) +
    geom_point() +
    geom_point(data = . %>% filter(fish_id1 == fish_id2), col = 'red') +
    #coord_fixed(1) +
    theme_minimal()
)
comp + aes(r_SNP_KING)
comp + aes(r_kmer)
comp + aes(r_kmer, r_SNP)
comp + aes(r_kmer, r_SNP_KING)

# There are clearly some individuals where the pedigree doesn't match the outcome,
# we should have a closer look
ind <- filter(rel_joined, r_SNP > 0.1, r_pedigree == 0) %>% pull(id1)
ggplot2::`%+%`(comp, filter(rel_joined, id1 %in% ind)) + facet_wrap(vars(id1)) + theme_bw()
# From this, I conclude that sample 280 and sample 355 are not correct. This is because there is no
# correlation between the relatedness estimates, even for cousins etc.

# Here is the plot again, now without the wrong samples:
ggplot2::`%+%`(comp, filter(rel_joined, !(sample_id1 %in% c(280, 355)), !(sample_id2 %in% c(280, 355))))
# and combined, with the correct estimates in green.
comp + geom_point(
  data = filter(rel_joined, !(sample_id1 %in% c(280, 355)), !(sample_id2 %in% c(280, 355))),
  color = 'green'
)
# Much better!!
